### PR TITLE
Add sklearn-style fit/transform/predict API on BalanceFrame

### LINKED ISSUE #51: [FEATURE] Make `balance` API closer to the sklearn syntax
The current workflow of just calling `adjust` might not be the easiest API for general Python/sklearn users.

### PR DESCRIPTION
**Pull request by neuralsorcerer**
- Closes #51

### CODE REVIEW

**Review by talgalili — CHANGES_REQUESTED** [review]
Amazing work.
Could you please update the ARCHITECTURE document, as well as update the quickstart tutorial with an example of how this should be used?

**Comment by talgalili** [comment]
Also, part of the point of this work is to enable the fitting of the model on a sample of observations and then use it on the rest.
Could you please add support to that with an example in the tutorial?
(Could be done in follow up PR)

**Comment by talgalili** [comment]
@neuralsorcerer just to say - great job iterating on this!
I know it's not easy...

**Comment by neuralsorcerer** [comment]
@talgalili finally done 😭🙏

**Comment by neuralsorcerer** [comment]
Also, will create a seperate PR for any missing coverage.

**Comment by talgalili** [comment]
Crazy work @neuralsorcerer !

I'll take a look, although I'll need someone other than me to review it - so will probably happen in the next 2 days (just because it's such a big deal of a PR)

GREAT JOB!

**Comment by talgalili** [comment]
@neuralsorcerer here is a detailed analysis of the diff from a stronger model (claude 4.6 instead of copilot).
It raises some issues (not major, but worth addressing). I read through them and they are worth adressing. Please read carefully, and fix - once done - please ping me for another round of review.




## `[balance][PR] Add sklearn-style fit/transform/predict API on BalanceFrame`

**PR:** [D100525447](https://www.internalfb.com/diff/D100525447) | **GitHub PR:** #402 | **Closes:** Issue #51
**Author:** Soumyadip Sarkar (external contributor) | **Status:** Needs Review
**CI:** Failing (cleanup artifacts job — likely not code-related)

---

## Executive Summary

This is a **substantial, well-engineered PR** (~1,400 lines across 7 files) that adds sklearn-style convenience methods (`fit`, `fit_transform`, `transform`, `predict`, `predict_weights`, `with_fitted_ipw_model`) to `BalanceFrame`. The code quality is high — good docstrings, comprehensive tests (566 lines), proper error messages, and careful handling of index alignment and caching. The IPW weighting method gains two new opt-in parameters (`store_fit_matrices`, `store_fit_metadata`) that store fit-time artifacts for downstream consumption.

**Overall recommendation: Accept with minor revisions.**

---

## 1. Architecture: How the New API Fits

### 1.1 New Method Chain

```
bf.fit(method="ipw")           # → adjust() + store artifacts
  .transform(on="both")        # → stored/recomputed model matrices
  .predict(on="target")        # → stored/recomputed propensity scores
  .predict_weights()           # → weights_from_link() on stored links
```

### 1.2 Train/Holdout Split

```
fitted = train_bf.fit(method="ipw")
scored = holdout_bf.with_fitted_ipw_model(fitted)
scored.predict(on="sample")    # recomputes using fitted estimator on holdout data
```

### 1.3 Method Mapping

| New method | Maps to | Behavior |
|---|---|---|
| `fit(method="ipw")` | `adjust(method="ipw")` | Alias, but defaults `store_fit_matrices=True` and `store_fit_metadata=True` |
| `fit_transform(...)` | `fit(...)` | Pure alias (since fitting and applying are one operation in balance) |
| `transform(on=...)` | Returns stored model matrices | Retrieves/recomputes IPW design matrices |
| `predict(on=..., output=...)` | Returns stored propensity predictions | Retrieves/recomputes IPW probabilities or log-odds |
| `predict_weights()` | `weights_from_link(...)` | Reproduces responder weights from stored link values |
| `with_fitted_ipw_model(fitted)` | Deep-copies fitted model to new BF | Enables train/holdout splitting |

### 1.4 New State Added to BalanceFrame

- `_attached_ipw_model: dict[str, Any] | None` — stores a model dict from another fitted BF (separate from `_adjustment_model`)
- `_require_ipw_model()` — checks `_adjustment_model` first, falls back to `_attached_ipw_model`
- `_compute_ipw_matrices_from_fit()` — recomputes matrices from scratch when holdout data differs from training data

### 1.5 Affected Files

| File | Change |
|---|---|
| `balance_frame.py` | +701 lines (new methods) |
| `weighting_methods/ipw.py` | +37 lines (`store_fit_matrices`, `store_fit_metadata`, model output) |
| `tests/test_balance_frame.py` | +566 lines (new test class) |
| `tests/test_ipw.py` | +63 lines (new tests for store_fit_* params) |
| `ARCHITECTURE.md` | Updated with fit-artifact workflow section |
| `CHANGELOG.md` | Updated with new API entry |
| `tutorials/balance_quickstart_new_api.ipynb` | New cells demonstrating the API |

---

## 2. Strengths

### 2.1 Immutability Preserved
`fit()` returns a new object, consistent with `adjust()`. `with_fitted_ipw_model()` deep-copies sample/target data. `fit()` with a `SampleFrame` target calls `set_target(in_place=False)` first.

### 2.2 Excellent Error Messages
Every failure path provides actionable guidance. Examples:
- `"Call BalanceFrame.fit(method='ipw') or run ipw(..., store_fit_matrices=True)"`
- `"fit(method='ipw', na_action='drop') is incompatible with stored fit artifacts because dropped rows break index/shape alignment"`

### 2.3 Smart Caching and Recomputation
`transform()`/`predict()` detect stale artifacts via index comparison and recompute on demand. Tests verify recomputation happens exactly once (via `unittest.mock.patch` call counting).

### 2.4 Mathematical Correctness Verified
- `predict_weights()` matches `adjusted.weight_series` within `rtol=1e-6`
- Holdout weights match independent `weights_from_link()` computation
- `predict(output="probability")` equals `expit(predict(output="link"))` via `np.testing.assert_allclose`
- Holdout uses its own design weights (not training weights) — tested with explicitly different weight values

### 2.5 Well-Designed IPW Parameters
- `store_fit_matrices` / `store_fit_metadata` default to `False` in `ipw()` (backward compat) but `True` in `fit()` (convenience)
- Clean separation: matrices for `transform()`, metadata for `predict()`
- Indices and weights are `.copy()`-ed, preventing aliasing bugs

### 2.6 No Data Leakage Risk
- Stored artifacts are from fit-time data
- `StandardScaler` uses `transform()` (not `fit_transform()`) on holdout data
- `na_action='drop'` incompatibility with stored artifacts is properly guarded

### 2.7 Comprehensive Tests (566 lines)
25+ test methods covering: basic flow, error paths, caching behavior, index alignment (unique, duplicate, reordered), holdout workflows, mathematical correctness, and edge cases.

### 2.8 Good Documentation
ARCHITECTURE.md, CHANGELOG.md, and tutorial notebook all properly updated with working examples.

---

## 3. Issues

### 3.1 HIGH: `_compute_ipw_matrices_from_fit` Duplicates `ipw()` Preprocessing

**Location:** `balance_frame.py:1201-1304`

The method reconstructs the full IPW preprocessing pipeline (~100 lines: transformations → model_matrix → scaler → penalties → type conversion) by reading stored model parameters. This parallels `ipw.py:743-907`.

**Risk:** Any future change to `ipw()` preprocessing must be mirrored here. No shared abstraction exists — just parallel implementations.

**Recommendation:** Extract shared preprocessing into a reusable function (e.g., `_build_ipw_design_matrix(sample_covars, target_covars, **config)`)

### 3.2 MEDIUM: `fit_transform()` Semantic Mismatch with sklearn

In sklearn, `fit_transform()` returns transformed data (`X_new`). Here it returns an adjusted `BalanceFrame` — identical to `fit()`, adding zero functionality. The docstring warns:

> Unlike sklearn's matrix-oriented `fit_transform`, this returns a fitted-and-adjusted BalanceFrame (it does not return X).

**Recommendation:**
- Make `fit_transform()` return `self.fit(...).transform(on="both")` for sklearn alignment

### 3.3 MEDIUM: `with_fitted_ipw_model()` Shared Estimator Not Documented

```python
out._attached_ipw_model = dict(model)  # shallow copy
```

The sklearn estimator (`model["fit"]`) is **shared** between original and holdout objects. This is intentional (saves memory, tested with `assertIs`), but the docstring doesn't mention it. Also, no covariate schema validation is performed at attachment time — column mismatches only surface later during `transform()`/`predict()`.

**Recommendation:**
- Document shared estimator in docstring
- Consider adding column-name validation at attachment time

### 3.4 MEDIUM: `transform()`/`predict()` Silently Mutate the Model Dict

These methods cache recomputed artifacts by modifying the model dict (e.g., `model["model_matrix_sample"] = sample_matrix`). This is a side effect in what looks like a read-only operation.

**Recommendation:** Document caching behavior and memory implications in docstrings.

### 3.5 LOW: Polymorphic Return Types

`transform(on="both")` returns `tuple[DataFrame, DataFrame]`, while `transform(on="sample")` returns `DataFrame`. This type instability can cause bugs.

**MAYBE:** Consider convenience methods `transform_sample()` / `transform_target()`? Something else?

### 3.6 LOW: `_align_frame_to_index` / `_align_series_to_index` Duplication

Nearly identical ~25-line methods for DataFrames and Series. Could share a generic helper.

### 3.7 LOW: `predict_weights()` Silent Fallback

When stored training weights don't match current data, `predict_weights()` silently falls back to current weights. This is correct for holdout workflows but could mask issues.

**Recommendation:** Add a `logger.debug()` message when falling back.


---

## 4. API Design Assessment

### 4.1 Deviation from sklearn Conventions

| sklearn convention | This PR | Assessment |
|---|---|---|
| `fit()` mutates self, returns self | Returns new object | **Good** — matches BalanceFrame's immutable pattern |
| `transform()` takes `X` as argument | No argument; uses stored data | **Acceptable** — domain-specific |
| `predict()` returns class labels | Returns probabilities/link | **Acceptable** — domain-specific |
| `fit_transform()` returns transformed `X` | Returns BalanceFrame | **Confusing** — see issue 3.2 |

### 4.2 The `on` Parameter Pattern

Using `on="sample"/"target"/"both"` is pragmatic for the survey weighting domain. It doesn't exist in sklearn but maps naturally.

### 4.3 IPW-Only Limitation

`transform`/`predict`/`predict_weights` only work with IPW. This is reasonable because:
- IPW is the only method producing a transferable propensity model with `predict_proba` semantics
- Other methods (rake, poststratify, cbps) work through different mathematical frameworks
- Error messages clearly explain the limitation
- `fit()` itself works with all methods — only downstream methods are IPW-restricted

---

## 5. Missing Test Cases

### 5.1 Medium Priority

| Gap | Description |
|---|---|
| NaN covariates | No test with actual missing values flowing through `transform()`/`predict()` with `na_action="add_indicator"` |
| Compound adjustment + sklearn API | `bf.fit("ipw").adjust("rake").predict()` — would the IPW model survive a second adjustment? |
| ASMD improvement verification | No sklearn-API test verifies that the weights actually *reduce bias* |
| Sum-of-weights invariant | No test checks that adjusted weights sum to the target population size |
| `fit()` with explicit `store_fit_*=False` | No test verifies that `fit(method="ipw", store_fit_matrices=False)` works but `transform()` raises |

### 5.2 Low Priority

| Gap | Description |
|---|---|
| `transform(on="both")` on holdout | Holdout tests only use `on="sample"` for `transform()` |
| `BalanceFrame` as `target` in `fit()` | Only `SampleFrame` and `None` targets tested |
| `with_fitted_ipw_model` with non-IPW model | No test passes a cbps-adjusted BF — should produce clear error |
| Degenerate inputs | No tests for empty DataFrames, single-row inputs, or zero-variance covariates |
| Extreme propensity scores | Near-0 or near-1 probabilities not tested |
| `transform()` self-consistency | No test verifies that `transform()` on self-fitted BF returns the exact stored fit-time matrix |
| Non-trivial `weight_trimming_percentile` | All tests use default trimming; winsorization path untested |

---

## 6. Test Quality Notes

### 6.1 Strengths
- Tests verify *mathematical correctness*, not just non-crashing
- Behavioral contracts are properly tested (immutability, caching counts, keyword-only enforcement)
- Good variety: unique indices, duplicate indices, reordered indices, holdout vs self-fitted

### 6.2 Concerns
- **Some tests are overly long** (e.g., `test_fit_on_subset_and_apply_to_holdout` tests 6 distinct behaviors). Could be decomposed.
- **Repetitive holdout setup** across many tests could benefit from a shared helper/fixture
- **Internal API testing** (directly mutating `model["sample_index"]`, popping keys) is fragile but appropriate — should be clearly marked
- **8-observation datasets** are adequate for functional tests but won't expose numerical instabilities (multicollinearity, near-separation)
- **No parametrized tests** despite highly repetitive patterns

---

## 7. IPW Changes Analysis (`ipw.py`)

### 7.1 New Parameters

```python
def ipw(..., store_fit_matrices: bool = False, store_fit_metadata: bool = False, ...) -> Dict[str, Any]
```

### 7.2 What Gets Stored

**`store_fit_metadata=True`** adds to model dict:
- `sample_link`, `target_link` — log-odds for sample/target rows
- `sample_probability`, `target_probability` — P(class=1) for sample/target rows
- `sample_index`, `target_index` — index copies
- `fit_sample_weights`, `fit_target_weights` — design weight copies

**`store_fit_matrices=True`** adds to model dict:
- `model_matrix_sample`, `model_matrix_target` — the X matrices used during fitting
- `sample_index`, `target_index` — index copies

### 7.3 Always Stored (regardless of flags)

New keys added to model output for downstream reconstruction:
- `fit_scaler` — StandardScaler instance (or None)
- `fit_penalties_skl` — penalty factor list (or None)
- `fit_matrix_type` — "sparse", "dense", or "dataframe"

### 7.4 Assessment

The design is sound:
- Opt-in with sensible defaults (False in ipw(), True in fit())
- Copies are made for indices and weights (preventing aliasing)
- No data leakage risk
- Redundant `sample_index`/`target_index` storage when both flags are True is harmless

---

## 8. Action Items

Fix all issues mentioned.

**Comment by neuralsorcerer** [comment]
@talgalili done :)

**Comment by talgalili** [comment]
GREAT - I'll take a look.

**Comment by talgalili** [comment]
This PR was split, updated, and published here:
https://github.com/facebookresearch/balance/pull/410
